pallets/marketplace/src/lib.rs

```diff
@@ -213,24 +255,109 @@ pub mod pallet {
         pub fn delete_advertisement(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
             let who = ensure_signed(origin)?;
 
-            let ad =
-                <StoredAdvertisement<T>>::get(&who).ok_or(Error::<T>::AdvertisementNotFound)?;
+            <StoredAdvertisementRestriction<T>>::get(&who)
+                .ok_or(Error::<T>::AdvertisementNotFound)?;
 
-            // update index
-            for pricing in &ad.pricing {
-                <StoredAdIndex<T>>::mutate(&pricing.reward_asset, |option_ads| {
-                    option_ads
-                        .as_mut()
-                        .map(|ads| ads.retain(|v| v.0 != who.clone()));
-                });
-            }
+            // prohibit updates as long as jobs assigned
+            ensure!(
+                !Self::has_matches(&who),
+                Error::<T>::CannotDeleteAdvertisementWhileMatched
+            );
 
-            <StoredAdvertisement<T>>::remove(&who);
+            let _ = <StoredAdvertisementPricing<T>>::clear_prefix(&who, MAX_PRICING_VARIANTS, None);
```

Comment:
Not sure what this is doing. Please add a comment to describe it.

Reply:
It's clearing all advertisements pricing belonging to the advertisement. I think the clear_prefix docs explain well what the arguments mean.